packages/xl-ai/src/api/formats/base-tools/createAddBlocksTool.ts

```diff
@@ -0,0 +1,290 @@
+   *   (the projection should be the the BlockNote document without the unsupported features)
+   */
+  rebaseTool: (
+    id: string,
+    editor: BlockNoteEditor<any, any, any>,
+  ) => Promise<RebaseTool>;
+  /**
+   * Converts the operation from `AddBlocksToolCall<T>` to `AddBlocksToolCall<PartialBlock<any, any, any>>`
+   *
+   * When using these factories to create a tool for a different format (e.g.: HTML / MD),
+   * the `toJSONToolCall` function is used to convert the operation to a format that we can execute
+   */
+  toJSONToolCall: (
+    editor: BlockNoteEditor<any, any, any>,
+    chunk: {
+      operation: AddBlocksToolCall<T>;
+      isUpdateToPreviousOperation: boolean;
+      isPossiblyPartial: boolean;
+    },
+  ) => Promise<AddBlocksToolCall<PartialBlock<any, any, any>> | undefined>;
+}) {
+  return (
+    editor: BlockNoteEditor<any, any, any>,
+    options: {
+      idsSuffixed: boolean;
+      withDelays: boolean;
+      onBlockUpdate?: (blockId: string) => void;
+    },
+  ) => {
+    const schema =
+      typeof config.schema === "function"
+        ? config.schema(editor)
+        : config.schema;
+    return streamTool<AddBlocksToolCall<T>>({
+      name: "add",
+      description: config.description,
+      parameters: {
+        type: "object",
+        properties: {
+          referenceId: {
+            type: "string",
+            description: "MUST be an id of a block in the document",
+          },
+          position: {
+            type: "string",
+            enum: ["before", "after"],
+            description:
+              "`after` to add blocks AFTER (below) the block with `referenceId`, `before` to add the block BEFORE (above)",
+          },
+          blocks: {
+            items: schema.block,
+            type: "array",
+          },
+        },
+        required: ["referenceId", "position", "blocks"],
+        $defs: schema.$defs,
+      },
+      validate: (operation) => {
+        if (operation.type !== "add") {
+          return {
+            result: "invalid",
+            reason: "invalid operation type",
+          };
+        }
+
+        if (operation.position !== "before" && operation.position !== "after") {
+          return {
+            result: "invalid",
+            reason: "invalid position",
+          };
+        }
+
+        if (!operation.referenceId || !operation.blocks) {
+          return {
+            result: "invalid",
+            reason: "referenceId and blocks are required",
+          };
+        }
+
+        let referenceId = operation.referenceId;
+        if (options.idsSuffixed) {
+          if (!referenceId?.endsWith("$")) {
+            return {
+              result: "invalid",
+              reason: "referenceId must end with $",
+            };
+          }
+
+          referenceId = referenceId.slice(0, -1);
+        }
+
+        const block = editor.getBlock(referenceId);
```

Comment:
This is always pulling from the current editor state. I would probably have an in-memory version of the document & pull from that instead. Probably more of an issue for collab mode than anything